pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploymentchaincontroller
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	// syncDeploymentChainsInterval represents time to sync state for all controlling deployment chains.
+	syncDeploymentChainsInterval = 15 * time.Second
+	// syncUpdatersInterval represents time to update list of controlling deployment chain updaters.
+	syncUpdatersInterval = time.Minute
+	// updaterWorkerNum represents the maximum number of updaters which can be
+	// run at the same time.
+	maxUpdaterWorkerNum = 10
+)
+
+type DeploymentChainController struct {
+	applicationStore     datastore.ApplicationStore
+	deploymentStore      datastore.DeploymentStore
+	deploymentChainStore datastore.DeploymentChainStore
+	// Map from deployment chain ID to the updater
+	// who in charge for the deployment chain updating.
+	updaters map[string]*updater
+	// WaitGroup for waiting the fetcher and updaters to stop.
+	wg sync.WaitGroup
+
+	logger *zap.Logger
+}
+
+func NewDeploymentChainController(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *DeploymentChainController {
+	return &DeploymentChainController{
+		applicationStore:     datastore.NewApplicationStore(ds),
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	syncDeploymentChainsTicker := time.NewTicker(syncDeploymentChainsInterval)
+	defer syncDeploymentChainsTicker.Stop()
+	syncUpdatersTicker := time.NewTicker(syncUpdatersInterval)
+	defer syncUpdatersTicker.Stop()
+
+	d.logger.Info("start running deployment chain controller")
+
+	d.wg.Add(1)
```

Comment:
Let's use "golang.org/x/sync/errgroup" to be simple.
Like this https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/cmd/piped/piped.go#L344

Reply:
Actually, we don't need 2 goroutines for those control loops. A `for` with 2 tickers is enough, I think.
In that way, we also don't have to care about locking on the `d.updaters` list since everything is done in a single goroutine.

``` golang
for {
  select{
  case <-ctx.Done():
  case <-syncUpdaterTicker.C:
  case <- syncChainsTicker.C:
}

Reply:
Get your point 👍 Lets me adopt your suggestion.

Reply:
On the second thought, since those 2 tickers can be run overlapping so I guess we still need to implement lock for `d.updaters` 🤔

Reply:
I was wrong, for select ensure that only one ticker can be run at a time, no overlapping so we don't need implement lock here 🙏  fixed by https://github.com/pipe-cd/pipe/pull/2986/commits/4e9e813aeccce64c12d9fb6e6f769aa6602c5b44